### ISSUE TITLE
Swallowed exception 'AmqpTimeoutException: No available channels'

### ISSUE THREAD

**Issue by delta11** [spring-projects/spring-amqp#2595]
**In what version(s) of Spring AMQP are you seeing this issue?**

2.4.13 & 2.4.17 & 3.1.1

**Describe the bug**

TLDR; under specific configs `AmqpTimeoutException: No available channels` exception is swallowed.

When creating more consumers than channels available, the start up becomes slow before eventually timing out.

Below is a stacktrace of where the threads start hanging in a wait.

```
"dynamic-registered:customer.141225-b06cd9d7-2fd1-4ba8-9ccd-f26598b3a53a-185" #9814 prio=5 os_prio=0 cpu=0.09ms elapsed=4.26s tid=0x00007f127430a3f0 nid=0x229a waiting on condition  [0x00007f10b69e8000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@17.0.6/Native Method)
	- parking to wait for  <0x0000000702fade38> (a java.util.concurrent.Semaphore$NonfairSync)
	at java.util.concurrent.locks.LockSupport.parkNanos(java.base@17.0.6/LockSupport.java:252)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@17.0.6/AbstractQueuedSynchronizer.java:717)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(java.base@17.0.6/AbstractQueuedSynchronizer.java:1074)
	at java.util.concurrent.Semaphore.tryAcquire(java.base@17.0.6/Semaphore.java:415)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.obtainPermits(CachingConnectionFactory.java:561)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.getChannel(CachingConnectionFactory.java:528)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.access$1600(CachingConnectionFactory.java:100)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.createChannel(CachingConnectionFactory.java:1418)
	at org.springframework.amqp.rabbit.connection.ConnectionFactoryUtils$RabbitResourceFactory.createChannel(ConnectionFactoryUtils.java:337)
	at org.springframework.amqp.rabbit.connection.ConnectionFactoryUtils.doGetTransactionalResourceHolder(ConnectionFactoryUtils.java:151)
	at org.springframework.amqp.rabbit.connection.ConnectionFactoryUtils.getTransactionalResourceHolder(ConnectionFactoryUtils.java:97)
	at org.springframework.amqp.rabbit.connection.ConnectionFactoryUtils.getTransactionalResourceHolder(ConnectionFactoryUtils.java:80)
	at org.springframework.amqp.rabbit.listener.BlockingQueueConsumer.start(BlockingQueueConsumer.java:613)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.initialize(SimpleMessageListenerContainer.java:1385)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.run(SimpleMessageListenerContainer.java:1230)
	at java.lang.Thread.run(java.base@17.0.6/Thread.java:833)
```

Eventually this wait will throw an exception and unwind the stack, which will get caught in `SimpleMessageListenerContainer` around line `1310` by the below catch block which will decide to swallow it preventing users from getting any logging info as to what's going on.

```java
catch (Throwable t) { //NOSONAR
  // by now, it must be an exception
  if (isActive()) {
	  logConsumerException(t);
  }
}
```

I'd like to discuss what the proper fix would be here if I'm understanding this all correctly. It feels like either a specialized catch clause is missing here, or this was writing either not knowing about the possibility of a `AmqpTimeoutException` or assuming it inherits from an exception like `AmqpIOException` so that it's already being handled correctly.

**To Reproduce**

I've attached a sample repo, running that will cause the exception to happen from 2 different paths.
Once during checking the declaration of the queues, which works as expected and logs a clear message.
Then there's a slightly more rare case of it happening from with a stack similiar to the one shown above which swallows the exception. That's the problem case.

**Expected behavior**

Logging a message regarding `org.springframework.amqp.AmqpTimeoutException: No available channels`

**Sample**

https://github.com/delta11/spring-amqp-channel-timeout-demo

**Comment by artembilan** [comment]
What you show is a thread dump and it really doesn't give any logging messages.

Since we constantly trying to redeclared queues, we really cannot release permits for consumer to be started and call that ` logConsumerException(t);` respectively.

Please, take a look into that method logic to see if you can get some clue.
For example, pay attention towards the end:
```
				else if (logger.isWarnEnabled()) {
					logger.warn(
							"Consumer raised exception, processing can restart if the connection factory supports it. "
									+ "Exception summary: " + t);
				}
			}
			publishConsumerFailedEvent("Consumer raised exception, attempting restart", false, t);
```
What I think you trying to point exactly to that `warn` message, with does swallow a stack trace, but at the same time we publish a `ListenerContainerConsumerFailedEvent` with the whole exception.
You can listen for that event and react respectively.

**Comment by delta11** [comment]
Thank you for the fast reaction! 

I missed that part, I tried it out and I can listen to that event. But in my example and my normal code base I don't get that warn line. It picks the debug line in the other case.
```java
logger.debug(
  "Consumer raised exception, processing can restart if the connection factory supports it",
t);
```
Maybe I don't understand the context, but I feel like running out of channels deserves more than a debug log line. What do you think?

**Comment by artembilan** [comment]
Well, this `No available channels` is a temporary state of the system. Typically after some operation the channel s going to come to the pool. Therefore we treat it as an interim, retryable error.

We may consider to raise some fatal error after a number of attempts talking to the pool, but that is already an improvement and it won't make it down as a back-port, especially to those versions which are already out of support: https://spring.io/projects/spring-amqp/#support

**Comment by delta11** [comment]
Clear, thank you for the context!
